### PR TITLE
Fixes #263: commit watcher is added to track files in every commit

### LINKED ISSUE #263: generate documentation only if any changes in documentation
### This is a
- [ ] Bug report
- [x] Feature request

### Current behaviour

documentation is generated for each and every commit in CI even if there are no changes in documentation files which wasting resources
### Expected behavior

generate documentation only if any changes happens to documentation files and also it enhances the efficiency of yaydoc

### PR DESCRIPTION
**Pull request by sch00lb0y**
## Description

I have used GitHub APIs to check what files has been changed in the last commit and by making use of this I'm generating documentation.
## Related Issue

close #263 
## Motivation and Context

now documentation is being generated for each on every commit even if there are no changes in documentation files so it wastes resources. by having this feature we able to generate and push documentation only if any changes in the documentation.
## Types of changes

- [ ] Bug fix 
- [x] New feature

## Checklist:

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] There is a corresponding issue for this pull request.
- [x] Mentioned the Issue number in the pull request commit message `Fixes #<number> commit message`
- [x] There is only one commit per issue.

### CODE REVIEW

**Comment by sch00lb0y** [comment]
deployment url https://demo08.herokuapp.com/

File: routes/ci.js [hunk 1]
```diff
@@ -73,38 +74,58 @@ router.post('/webhook', function(req, res, next) {
...
-                email: result.email,
-                gitURL: req.body.repository.clone_url,
-                username: result.username,
-                uniqueId: generatedData.uniqueId,
-                encryptedToken: result.accessToken
-              })
+      let sha = req.body.head_commit.id;
+      let username = req.body.repository.full_name.split("/")[0];
+      let repoName = req.body.repository.name;
+      let formats = ['.md', '.rst'];
```
Comment (imujjwal96) [inline routes/ci.js]: what will happend when i include any external file, for example an image to the repository?

Comment (sch00lb0y) [inline routes/ci.js]: It won't generate the documentation. if you want me to add some more formats. I'll add it to the list.

**Comment by imujjwal96** [comment]
What will happen when i make changes to my api docs?

**Comment by sch00lb0y** [comment]
right now we having support to generate documentation from .rst and .md. As soon as we implement swagger, api blueprint. we'll add those format to the list.

**Comment by imujjwal96** [comment]
What about javadocs?

**Comment by pri22296** [comment]
@imujjwal96 Good point about javadocs. For generating any kind of API documentation we'll need the source files as well. Including all of those extensions will kind of include the entire repository making this PR pointless.

**Comment by sch00lb0y** [comment]
@imujjwal96  Thanks for mentioning.  This we can solve by downloading the .yaydoc.yml before hand itself. I'll check whether they have a config for auto api, Javadoc. if it there I'll generate documentation as normal flow. If not I'll check for any changes in '.md' or '.rst'. @pri22296 @imujjwal96 your view on this solution?

**Comment by pri22296** [comment]
I'm just thinking if doing all that(even then there is a chance that we may miss some files which affect the documentation's final look) is really worth it. I mean you would have to include the extension `yml` as the user can modify just the `.yaydoc.yml` file. If the user provides a custom theme via the _theme directory, they may modify the `html` file or the `theme.conf` file. So you would have to include them to. Hell, the user could just upload a new version of a custom theme to PyPI which would require no change in their own repository thus the build won't run. I'm not sure how would all those cases be handled?

**Comment by sch00lb0y** [comment]
@pri22296  your point so true. what we can do is. if any changes in doc path, if they any external config like Javadoc, auto API. we'll generate documentation. otherwise not. 
>Hell, the user could just upload a new version of a custom theme to PyPI which would require no change in their own repository thus the build won't run.

This is exceptional (I know this is important. for this we can mention in Readme.md do manual generation for that one time)

**Comment by pri22296** [comment]
> manual generation for that one time

So we force the users to manually generate when we can handle that situation automatically?

Technically, users can also include arbitrary text files with the include directive. This feature is used to include documents such as `LICENSE` which usually are without extension.

Optimization is a good thing, I understand that. But removing functionalities for optimization may not be a good idea here. 

Let's ask for suggestions from the mentors. @niranjan94 your views?

**Comment by pri22296** [comment]
@sch00lb0y I don't think this would be needed or implemented anytime soon. close this?

**Pull request closed without merging**